_posts/2021-01-18-smtchecker-and-synthesis-of-external-functions.md

```diff
@@ -0,0 +1,420 @@
+
+Another important feature is the one that names this post. In case *it is*
+possible to violate the assertion, the solver synthesizes the behavior of the
+externally called unknown code, telling us exactly what needs to be done to
+break the property. If `Mutex.run` is not guarded by `lock`, the SMTChecker
+reports
+
+```bash
+Warning: CHC: Assertion violation happens here.
+Counterexample:
+x = 1, lock = false, unknown = 1
+
+Transaction trace:
+Mutex.constructor(1)
+State: x = 0, lock = false, unknown = 1
+Mutex.run()
+    unknown.run() -- untrusted external call, synthesized as:
+        Mutex.set(1) -- reentrant call
+  --> mutex.sol:34:3:
+   |
+34 | 		assert(xPre == x);
+   | 		^^^^^^^^^^^^^^^^^
+```
+
+Here the solver finds out that all the unknown code needs to do is call
+`Mutex.set` once, and the property will be violated at the end of `Mutex.run`.
+
+Note that `unknown = 1` means the address of the `Unknown` contract called by
+`Mutex`, where the actual address of the deployed contract doesn't really
+matter, as long as it implements a function `run()` that calls `Mutex.set(1)`.
+
+---
+
+The following sample contract shows the counterexample report in case the external
+unknown function needs to make multiple calls:
+
+```solidity
+//SPDX-License-Identifier: GPL-v3
+pragma solidity ^0.8.0;
+pragma experimental SMTChecker;
+
+interface Unknown {
+	function run() external;
+}
+
+contract Test {
+	uint x;
+	uint y;
+
+	Unknown immutable unknown;
+
+	constructor(Unknown _u) {
+		require(address(_u) != address(0));
+		unknown = _u;
+	}
+
+	function incX() public { ++x; }
+	function incY() public { ++y; }
+
+	function f() public {
+		unknown.run();
+		assert(x < 2 || y < 2);
+	}
+}
+```
+
+The property is false, and the SMTChecker reports:
+
+```bash
+$ solc external_inc.sol --model-checker-engine chc
+
+Warning: CHC: Assertion violation happens here.
+Counterexample:
+x = 3, y = 1, unknown = 1
+
+Transaction trace:
+Test.constructor(1)
+State: x = 0, y = 0, unknown = 1
+Test.f()
+    unknown.run() -- untrusted external call, synthesized as:
+        Test.incX() -- reentrant call
+        Test.incX() -- reentrant call
+        Test.incX() -- reentrant call
+        Test.incY() -- reentrant call
+  --> external_inc.sol:25:3:
+   |
+25 | 		assert(x < 3 || y < 1);
+   | 		^^^^^^^^^^^^^^^^^^^^^^
+```
+
+---
+
+We can naturally try to apply this feature on code that might
+be vulnerable to reentrancy. For example, take the following simplified/modified
+implementation of a [ERC777](https://eips.ethereum.org/EIPS/eip-777) token:
+
+```solidity
+//SPDX-License-Identifier: GPL-v3
+pragma solidity ^0.8.0;
+pragma experimental SMTChecker;
+
+interface IERC777Recipient {
+	function tokensReceived(
+		address from,
+		address to,
+		uint256 amount
+	) external;
+}
+
+interface IERC777Sender {
+	function tokensToSend(
+		address from,
+		address to,
+		uint256 amount
+	) external;
+}
+
+contract ERC777 {
+	mapping (address => uint256) balances;
```

Comment:
Would be a little closer to the real `ERC777` if this was called `balanceOf` and made `public`.